### PR TITLE
controller helpers

### PR DESCRIPTION
**Pull request by henderiw**
this is a set of controller helpers to make it easier to build controllers

### CODE REVIEW

File: controllers/pkg/resource/api.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// NewAPIPatchingApplicator returns an Applicator that applies changes to an
+// object by either creating or patching it in a Kubernetes API server.
+func NewAPIPatchingApplicator(c client.Client) APIPatchingApplicator {
+	return APIPatchingApplicator{c}
+}
+
+// Apply changes to the supplied object. The object will be created if it does
+// not exist, or patched if it does. If the object does exist, it will only be
+// patched if the passed object has the same or an empty resource version.
+func (a *APIPatchingApplicator) Apply(ctx context.Context, o client.Object, ao ...ApplyOption) error {
```
Comment (aakashchan) [inline controllers/pkg/resource/api.go:46]: `sigs.k8s.io/cluster-api/util/patch` package have PatchHelper struct. Is there a reason why we don't want to use it?

**Comment by aakashchan** [comment]
Do we need a helper on the controller-runtime utils for Finalizer and Patching?

File: controllers/pkg/resource/api.go (hunk 1 above)
Comment (henderiw) [reply controllers/pkg/resource/api.go:46]: The whole idea is to present the client the same semantics as the regular client. E.g. Apply and Create  for the developer have the exact same semantics. So under the hood patch is used.

if err := r.Apply(ctx, secret); err != nil {
			cr.SetConditions(infrav1alpha1.Failed(err.Error()))
			r.l.Error(err, "cannot create secret")
			return err
		}

**Comment by henderiw** [comment]
eventually I would like to appear in controller runtime, but I did not find it anywhere so far in a way that it extends the client gracefully.

**Comment by gvbalaji** [comment]
This may have been covered already by the bootstrap controller PR? @henderiw

**Comment by henderiw** [comment]
this is covered by the bootstrap controller now

**Pull request closed without merging**